### ISSUE TITLE
Parallel torrent verification

### ISSUE THREAD

**Issue by Avamander** [transmission/transmission#5064]
### What is the issue?

Verifying a significant amount of data one-torrent-by-one can take a long long time. 

Ideally data verification would have some amount of per-torrent parallelism. Even one additional parallel verification would very likely waste less time - for example two queues, one starting from the largest and other from the smallest amount of data. 

A solution approximately like this would allow smaller torrents to resume/complete sooner, while the larger ones are also eventually churned through.

### Which application of Transmission?

transmission-daemon

### Which version of Transmission?

4.0.1

**Comment by ckerr** [comment]
I'm not sure this would actually make things faster?

Generally the bottleneck is disk IO, so verifying multiple torrents in parallel should make them all verify slower.

Also, the verification queue already places smaller torrents closer to the front of the queue, so going single-file already benefits small files.

**Comment by killemov** [comment]
@ckerr Correct, so parallelism would hurt same-disk verification and benefit verification across multiple disks.

**Comment by ckerr** [comment]
Next question is whether there's a reliable + portable way to determine what physical disk a torrent's files are on.

Sometimes in FLOSS the best way to find out how to do something is to say it can't be done so that someone will come out of nowhere and prove you wrong, so... 

## IMO simple, reliable, and portable code doesn't exist to determine what physical disk a file is on.

**Comment by mikedld** [comment]
Aren't we already doing it in `tr_sys_path_is_same()` (`st_dev` in file-posix.cc and `dwVolumeSerialNumber` in file-win32.cc)? We already rely on it => reliable, we already have a function for it (kinda) => portable :-P

**Comment by mikedld** [comment]
Ah, sorry, those are logical volumes :(

**Comment by mikedld** [comment]
I'm afraid branching on physical location might be ineffective if e.g. you have some sort of RAID controller (not a physical disk itself) which is super-fast (can handle validating 3 torrents concurrently) and spans multiple disks (which one contains the file?). If I/O is a bottleneck, could be just measuring how fast things go (per logical volume?) might be a better option? And then, if a torrent has files on different disks/volumes (e.g. due to symlinks), do we serialize? Sounds kinda complicated...

**Comment by Avamander** [comment]
Better storage awareness would be nice, but how certain is it that the disk IO is the bottleneck (over say CPU or disk IOPS)? Maybe it's just the sample rate of iotop, but to me it seems like there are dips in RW each time when initializing-verifying-completing a bunch of small-sized torrents. Is it very unlikely that parallelism would provide better/higher average disk IO utilization? (Thus being faster in the end)

**Comment by killemov** [comment]
Mkay, adding a simple configurable number for concurrent threads it is then. Also the concept of root paths could be introduced and they could be be grouped in tiers that cannot do multiple verification within themselves.

**Comment by ckerr** [comment]
No, I'm not going to add a configuration field for the number of verification threads; this doesn't warrant [Yet Another Setting](https://ometer.com/preferences.html).

If someone has a way to determine what physical disk a file is on then that's useful in every scenario without a new setting & I can work with that.

**Comment by ilikepeaches** [comment]
Related: if several torrents share data & hash (e.g. torrents from several separate private trackers), they should verify ~together rather than separately/sequentially, no?

**Comment by Tharn** [comment]
Maybe this is too simple an idea, but Transmission could start one file verify thread for each destination folder and then iterate through these to see which ones are choking each other for speed.

**Comment by ckerr** [comment]
Too much guesswork IMO. You'd have to match all the folders to guess which ones are on the same partition and then there's always the possibility that some other process is causing IO on the same disk, skewing results. I don't see a good way of doing this without knowing which physical disk a file is on.

**Comment by rekh127** [comment]
I don't think you can generalize that disk I/O is the bottleneck.  We have 'disks' in common useage from 180 MBps to 7500 MBps, and thats for a single drive, excluding RAID arrays which many people could be running on.   

And CPU's are just as widespread, are they running on a raspberry pi, writing to a NAS with a SSD raid array? or are they running on a massive recent ryzen and write to a single HDD they use for their linux iso's. 

I can say on my server a verification caps out a thread and is only reading ~250 MBps. I know the storage can handle a lot more mixed read/write than that.


Seems like making a configurable number of threads allows people to adjust to their situation. Sure someone might slow themselves down by running too many on a single HDD and cause too much seeking. But thats on them.

**Comment by theDDA** [comment]
FWIW this thread is the final drop for me. I've tried to like Transmission, I really did, but between issues like this one, #2462 (and a related decade old #108) that are known for years, and the general mess this project seems to be in (maintainers going awol, PRs not being merged, the general attitude of the maintainer above) my impression is that not only it's not competitive - it's simply nonviable for serious use, and it's not going to be any time soon.

Like @duskmoss said above, an SSD or a NAS can reach multigigabyte speeds, it seems quite silly to limit yourself to single core hashing performance. Not being able to spawn a worker for that, or for _moving a file_ for over a decade is not a great look.

**Comment by spanevin** [comment]
There is at least one case not related to physical storage where parallel verification can make torrents downloading much faster.

We know for sure, that transmission can download data AND verify data in parallel, right? As when I add a new updated torrent, previous downloads don't pause until verification of the newly added one is done...

Another case - I'm adding an update for an existing torrent, which requires verification of some gigabytes of data, and then I'm adding a few NEW torrents which have no downloaded data so they could be downloaded in parallel with verification, but their download doesn't start until verification of the updated torrent is done...

Parallel verification could solve the issue. But it can be resolved in another way as well. Just check if there's some data to verify or not, and don't do waiting for torrents, which don't need verification...
